File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/NavigationFasterRouteListener.java
package com.mapbox.services.android.navigation.v5.navigation;

import androidx.annotation.Nullable;

import com.mapbox.api.directions.v5.models.DirectionsResponse;
import com.mapbox.services.android.navigation.v5.route.FasterRoute;
import com.mapbox.services.android.navigation.v5.route.RouteListener;
import com.mapbox.services.android.navigation.v5.routeprogress.RouteProgress;

import timber.log.Timber;

class NavigationFasterRouteListener implements RouteListener {

  private static final int FIRST_ROUTE = 0;

  private final NavigationEventDispatcher eventDispatcher;
  private final FasterRoute fasterRouteEngine;

  NavigationFasterRouteListener(NavigationEventDispatcher eventDispatcher, FasterRoute fasterRouteEngine) {
    this.eventDispatcher = eventDispatcher;
    this.fasterRouteEngine = fasterRouteEngine;
  }

  @Override
  public void onResponseReceived(DirectionsResponse response, @Nullable RouteProgress routeProgress) {
    if (fasterRouteEngine.isFasterRoute(response, routeProgress)) {
      eventDispatcher.onFasterRouteEvent(response.routes().get(FIRST_ROUTE));
    }
  }

  @Override
  public void onErrorReceived(Throwable throwable) {
    Timber.e(throwable);
  }
}

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/internal/navigation/NavigationFasterRouteListener.kt
package com.mapbox.services.android.navigation.v5.internal.navigation

import com.mapbox.api.directions.v5.models.DirectionsResponse
import com.mapbox.navigation.base.logger.model.Message
import com.mapbox.navigation.logger.MapboxLogger
import com.mapbox.navigation.utils.extensions.ifNonNull
import com.mapbox.services.android.navigation.v5.route.FasterRoute
import com.mapbox.services.android.navigation.v5.route.RouteListener
import com.mapbox.services.android.navigation.v5.routeprogress.RouteProgress

internal class NavigationFasterRouteListener(
    private val navigationEventDispatcher: NavigationEventDispatcher,
    private val fasterRouteEngine: FasterRoute
) : RouteListener {

    companion object {
        private const val FIRST_ROUTE = 0
    }

    override fun onResponseReceived(response: DirectionsResponse, routeProgress: RouteProgress?) {
        ifNonNull(routeProgress) {
            if (fasterRouteEngine.isFasterRoute(response, it)) {
                navigationEventDispatcher.onFasterRouteEvent(response.routes()[FIRST_ROUTE])
            }
        }
    }

    override fun onErrorReceived(throwable: Throwable) {
        MapboxLogger.e(Message(throwable.localizedMessage), throwable)
    }
}
